### PR TITLE
Fix projection in IPC reader

### LINKED ISSUE #1735: IPC reader may break on projection
**Describe the bug**

Function `read_record_batch` deals with `projection` in `arrow/src/ipc/reader.rs`.

Current logic may not advance the `node_index` and `buffer_index` correctly because it does not call `create_array` for skipped nodes.

**To Reproduce**
```
[dependencies]
arrow = "14.0.0"
```

```
use std::fs::File;
use std::sync::Arc;

use arrow::array::*;
use arrow::datatypes::*;
use arrow::ipc::reader::*;
use arrow::ipc::writer::*;
use arrow::record_batch::RecordBatch;

fn main() {
    let schema = Schema::new(vec![
        Field::new("f0", DataType::UInt32, false),
        Field::new("f1", DataType::Utf8, false),
        Field::new("f2", DataType::Boolean, false),
    ]);
    
    let array0 = UInt32Array::from(vec![1, 2, 3]);
    let array1 = StringArray::from(vec!["foo", "bar", "baz"]);
    let array2 = BooleanArray::from(vec![true, false, true]);

    let record_batch = RecordBatch::try_new(
        Arc::new(schema.clone()),
        vec![Arc::new(array0), Arc::new(array1), Arc::new(array2)]
    ).unwrap();

    {
        let file = File::create("./test.arrow_file").unwrap();
        let mut writer = FileWriter::try_new(file, &schema).unwrap();

        writer.write(&record_batch).unwrap();
        writer.finish().unwrap();
    }

    let projection = vec![1];
    let file = File::open("./test.arrow_file").unwrap();
    let arrow_reader = FileReader::try_new(file, Some(projection.clone()));
    let batch = arrow_reader.unwrap().next();
    println!("{:?}", batch);
}
```

Above code will produce:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidArgumentError("First offset 1 in Utf8 is smaller than last offset 0")', /home/yupeng/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow-14.0.0/src/ipc/reader.rs:287:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Expected behavior**
Output:

```
Some(Ok(RecordBatch { schema: Schema { fields: [Field { name: "f1", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None }], metadata: {} }, columns: [StringArray
[
  "foo",
  "bar",
  "baz",
]], row_count: 3 }))
```

**Additional context**

### PR DESCRIPTION
**Pull request by iyupeng**
# Which issue does this PR close?

Closes #1735 .

# Rationale for this change
 Fix possible panic caused by projection in IPC reader.

 

# What changes are included in this PR?
Fix the function `read_record_batch` in `arrow/src/ipc/reader.rs`.

# Are there any user-facing changes?
No.

### CODE REVIEW

**Review by viirya — COMMENTED** [review]
Thanks @iyupeng. Can you add a related test with the change? I saw you already have reproducible example in the issue. Seems it is easy to make it as a test?

**Comment by iyupeng** [comment]
Hi @viirya , thanks for your comment. A new test is added to IPC reader in my last commit.

File: arrow/src/ipc/reader.rs [hunk 1]
```diff
@@ -275,6 +275,120 @@ fn create_array(
     Ok((array, node_index, buffer_index))
 }
 
+/// Skip fields based on data types to advance `node_index` and `buffer_index`.
+/// This function should be called when doing projection in fn `read_record_batch`.
+/// The advancement logic references fn `create_array`.
+fn skip_filed(
```
Comment (viirya) [inline arrow/src/ipc/reader.rs:281]: I think you mean:

```suggestion
fn skip_field(
```

**Review by viirya — APPROVED** [review]
Looks a correct fix to me. I checked and compared it with `create_array`. 

Although I think we can still improve the test. For example, projection on complex types such as struct, list, union, etc.

**Comment by iyupeng** [comment]
Hi @viirya , the test for projection has been improved to check more data types. It is ready for review now.

**Review by viirya — APPROVED** [review]
Thanks @iyupeng! The new test looks good to me. I will leave this open for a while and will merge in next day if no more comments from other possible reviewers.

Thanks for the fix.

**Comment by viirya** [comment]
Merged, thanks!

**Pull request merged**